sql/archive.sql

```diff
@@ -0,0 +1,33 @@
+SET statement_timeout=0;
+
+-- Move old contracts to archive tables
+DO $$
+DECLARE
+    contract record;
+    link_contract record;
+BEGIN
+    FOR contract IN SELECT * FROM cards WHERE (type='execute@1.0.0' OR type='session@1.0.0' OR type='external-event@1.0.0' OR active=false)
+        AND created_at < to_char(CURRENT_DATE - INTERVAL '3 months', 'YYYY-MM-DD')::date
+    LOOP
+        RAISE INFO 'Moving contract: [%] [%]', contract.type, contract.id;
+        FOR link_contract IN SELECT * FROM links2 WHERE fromid=contract.id OR toid=contract.id
+        LOOP
+            RAISE INFO 'Moving link: [%]', link_contract.id;
+
+            -- Move link record from links2 table
+            WITH moved_rows AS (
+                DELETE FROM links2 WHERE id=link_contract.id RETURNING *
+            ) INSERT INTO archive_links2 SELECT * FROM moved_rows;
+
+            -- Move link record from cards table
+            WITH moved_rows AS (
+                DELETE FROM cards WHERE id=link_contract.id RETURNING *
```

Comment:
Am I right in reading this as that contracts are duplicated between `links2` and `cards` with the same id?

Reply:
Yes, I believe this is so because querying on links using the `cards` table is too slow, so a separate "meta" table specifically for `link` contracts was created.

Reply:
@Page- @joshbwlng just to expand a bit more on this. `link` contracts define a bidirectional edge even though it has a `from` and `to` field. But links also have a name and an inverse name like `has attachment` and `is attached to`. So the query for a link is:

```sql
(link.data.from.id = <parent ID> AND link.name = <name>) OR
(link.data.to.id = <parent ID> AND link.data.inverseName = <name>)
```

Even with proper indexes, PG does not like that. It always uses a bitmap scan, which is kinda horrible here as contracts are large, there's toast and millions of contracts. Index or not, there's still a lot of IO required.

With a separate `links` table, the situation is better because rows are much smaller. Only some fields are duplicated from the full `link` contract. But i doesn't fundamentally fix the issue of bitmap scans. Each row still denotes a bidirectional link so the query above doesn't change much.

With `links2`, each row denotes a single direction of the bidirectional edge. Each `link` contract has *two* rows in the `links2` table associated with it. The query becomes a much simpler:

```sql
links2.from = <parent ID> AND links2.name = <name>
```

PG is happy to do an index-only search and each row was also shrunk in this new version. The improvements in query time and resource usage are dramatic.

Reply:
@Ereski Do you think we can improve any of the link queries I have currently in the archive script? In this case, I'm thinking we can just get all links for the contracts we're archiving, and not worry about the name of the link.